src/dos/dos_classes.cpp

```diff
@@ -308,6 +309,119 @@ bool DOS_PSP::SetNumFiles(uint16_t file_num)
 	return true;
 }
 
+static constexpr auto bytes_to_read = 1024;
+
+std::optional<std::string> DOS_PSP::GetEnvironmentValue(const std::string_view variable) const
+{
+	/* Walk through the internal environment and see for a match */
+	PhysPt env_read = PhysicalMake(GetEnvironment(), 0);
+
+	char env_string[bytes_to_read + 1];
+	if (variable.empty()) {
+		return {};
+	}
+
+	for (;;) {
+		MEM_StrCopy(env_read, env_string, bytes_to_read);
+		if (!env_string[0]) {
+			return {};
+		}
+		env_read += (PhysPt)(safe_strlen(env_string) + 1);
+		char* equal = strchr(env_string, '=');
+		if (!equal) {
+			continue;
+		}
+		/* replace the = with \0 to get the length */
+		*equal = '\0';
+		if (strlen(env_string) != variable.size()) {
+			continue;
+		}
+		if (!iequals(variable, env_string)) {
+			continue;
+		}
+
+		return env_string + variable.size() + sizeof('=');
+	}
+}
+
+std::vector<std::string> DOS_PSP::GetAllRawEnvironmentStrings() const
+{
+	std::vector<std::string> all_env_vars = {};
+
+	char env_string[bytes_to_read + 1];
+	PhysPt env_read = PhysicalMake(GetEnvironment(), 0);
+	for (;;) {
+		MEM_StrCopy(env_read, env_string, bytes_to_read);
+		if (!env_string[0]) {
+			return all_env_vars;
+		}
+		all_env_vars.emplace_back(env_string);
+		env_read += (PhysPt)(safe_strlen(env_string) + 1);
+	}
+}
+
+bool DOS_PSP::SetEnvironmentValue(std::string_view variable, std::string_view new_string)
+{
+	PhysPt env_read = PhysicalMake(GetEnvironment(), 0);
+
+	// Get size of environment.
+	DOS_MCB mcb(GetEnvironment() - 1);
+	uint16_t envsize = mcb.GetSize() * 16;
+
+	PhysPt env_write                   = env_read;
+	PhysPt env_write_start             = env_read;
+	char env_string[bytes_to_read + 1] = {0};
+	const auto entry_length            = variable.size();
+	do {
+		MEM_StrCopy(env_read, env_string, bytes_to_read);
+		if (!env_string[0]) {
+			break;
+		}
+		env_read += (PhysPt)(safe_strlen(env_string) + 1);
+		if (!strchr(env_string, '=')) {
+			continue; /* Remove corrupt entry? */
```

Comment:
I suggest to add a warning log about damaged environment. Preferably have some variable (like `bool already_warned`) to avoid flooding the output log.

Reply:
Was this set to `resolved` by mistake?